File: BaekJoon/재귀/곱셈.py
"""
author: kyungmin
date: 23.12.12
title: 곱셈
time: 30분
"""
import sys
input = sys.stdin.readline

A, B, C = map(int, input().rstrip().split())

def recursive(a, b, c):
    if b == 0:
        return 1
    
    tmp = recursive(a, b//2, c) % C
    tmp = (tmp * tmp) % c

    if b % 2 == 0:
        return tmp
    else:
        return (tmp * a) % c
    
print(recursive(A, B, C))

File: BaekjoonWithSwift/BaekjoonWithSwift/재귀/곱셈.swift
//
//  곱셈.swift
//  BaekjoonWithSwift
//
//  Created by kyungmin on 2023/12/12.
//

import Foundation

func 곱셈() {
    let input = readLine()!.split(separator: " ").compactMap { Int($0) }
    let A = input[0]
    let B = input[1]
    let C = input[2]
    print(recursive(cnt: B))
    
    func recursive(cnt: Int) -> Int {
        if cnt == 0 {
            return 1
        }
        
        var tmp = recursive(cnt: cnt / 2) % C
        tmp = (tmp * tmp) % C
        
        if cnt % 2 == 0 {
            return tmp
        } else {
            return (tmp * A) % C
        }
    }
}
